### PR TITLE
Refactor search results view into its own component

### PR DESCRIPTION
**Pull request by rakshakhegde**
This might look like a huge PR with 600+ lines of changes. But it's mostly files moved around for refactoring search view into its own component, so I'm assuming this shouldn't be a problem to review.

https://www.pivotaltracker.com/story/show/165276354

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
@rakshakhegde Can you rebase this branch? Looks like there are some conflicts.

**Comment by rakshakhegde** [comment]
@vinaysshenoy conflicts fixed

File: app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt [hunk 1]
```diff
@@ -38,196 +28,41 @@ class PatientSearchResultsControllerTest {
...
-  @Test
-  fun `when screen is created and no matching patients are available then the empty state should be shown`() {
-    val emptySearchResults = PatientSearchResults(
-        visitedCurrentFacility = emptyList(),
-        notVisitedCurrentFacility = emptyList()
-    )
-    whenever(patientRepository.search(any())).thenReturn(Observable.just(emptySearchResults))
-
-    uiEvents.onNext(PatientSearchResultsScreenCreated(PatientSearchResultsScreenKey("name")))
+  fun `when SearchResultClicked event is emitted open PatientSummaryScreen`() {
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt]: Rename this to `when search result is clicked, then patient summary must be opened`?

File: app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt [hunk 2]
```diff
@@ -38,196 +28,41 @@ class PatientSearchResultsControllerTest {
...
-            userClock = userClock
-        ),
-        PatientSearchResultRow(
-            searchResult = patientSearchResult2,
-            currentFacility = currentFacility,
-            phoneObfuscator = phoneObfuscator,
-            dateOfBirthFormat = dateOfBirthFormatter,
-            clock = utcClock,
-            userClock = userClock
+  fun `when RegisterNewPatient event is emitted open PatientSummaryScreen`() {
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt]: Rename this to `when register new patient is clicked, then patient entry screen must be opened`?

File: app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsScreen.kt [hunk 3]
```diff
@@ -61,7 +47,7 @@ class PatientSearchResultsScreen(context: Context, attrs: AttributeSet) : Relati
     setupScreen()
 
     Observable
-        .mergeArray(screenCreates(), newPatientClicks(), adapterUiEvents)
+        .mergeArray(screenCreates(), searchResultsView.upstreamUiEvents)
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsScreen.kt]: done

File: app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        .mergeArray(screenCreates(), newPatientClicks(), adapterUiEvents, downstreamUiEvents)
+        .observeOn(Schedulers.io())
+        .compose(controller)
+        .observeOn(AndroidSchedulers.mainThread())
+        .takeUntil(RxView.detaches(this))
+        .subscribe { it(this) }
+  }
+
+  private fun setupScreen() {
+    hideKeyboard()
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt]: done

File: app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    inflate(context, R.layout.search_results_view, this)
+    if (isInEditMode) {
+      return
+    }
+    TheActivity.component.inject(this)
+    setupScreen()
+
+    Observable
+        .mergeArray(screenCreates(), newPatientClicks(), adapterUiEvents, downstreamUiEvents)
+        .observeOn(Schedulers.io())
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt]: done

File: app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  lateinit var controller: SearchResultsViewController
+
+  @Inject
+  lateinit var utcClock: UtcClock
+
+  val downstreamUiEvents: Subject<UiEvent> = PublishSubject.create()
+  val upstreamUiEvents: Subject<UiEvent> = PublishSubject.create()
+
+  private val adapter = GroupAdapter<ViewHolder>()
+  private val adapterUiEvents: Subject<UiEvent> = PublishSubject.create()
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt]: done

File: app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsViewController.kt [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    private val userSession: UserSession,
+    private val facilityRepository: FacilityRepository,
+    private val phoneObfuscator: PhoneNumberObfuscator,
+    private val utcClock: UtcClock,
+    val userClock: UserClock,
+    @Named("date_for_search_results") private val dateOfBirthFormatter: DateTimeFormatter
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = events
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsViewController.kt]: done

File: app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt (hunk 1 above)
Comment (rakshakhegde) [reply app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt]: done

File: app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt (hunk 2 above)
Comment (rakshakhegde) [reply app/src/test/java/org/simple/clinic/search/results/PatientSearchResultsControllerTest.kt]: done

File: app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import org.simple.clinic.R
+import org.simple.clinic.activity.TheActivity
+import org.simple.clinic.router.screen.ScreenRouter
+import org.simple.clinic.util.UtcClock
+import org.simple.clinic.widgets.UiEvent
+import org.simple.clinic.widgets.hideKeyboard
+import org.simple.clinic.widgets.visibleOrGone
+import javax.inject.Inject
+
+class SearchResultsView(context: Context, attrs: AttributeSet) : RelativeLayout(context, attrs) {
```
[the remark this answers is not in the document]
Comment (rakshakhegde) [reply app/src/main/java/org/simple/clinic/searchresultsview/SearchResultsView.kt]: done

**Comment by vinaysshenoy** [comment]
@rakshakhegde Looks good. I've made a couple of small changes:

- In `PatientSearchResultsScreen`, I changed how the mapping from the `PatientSearchView` events to the `PatientSearcgResultsScreen` was done to be more consistent with the way we set up streams elsewhere.
- When we moved to `ReplayUntilScreenIsDestroyed`, we need to explicitly provide a `ScreenDestroyed` stream, which was missed. I added these. We really should figure out how to enforce providing this stream in some way.

**Comment by vinaysshenoy** [comment]
@rakshakhegde What was this last change you pushed?

**Pull request merged**